docs/architecture/004-reframe-style.md

```diff
@@ -0,0 +1,89 @@
+# 004: re-frame handlers
+
+Authors: @hukka
+
+## Handling UI events and side-effects
+
+re-frame architecture allows splitting all side-effects from event handlers.
+The input (not covered by function variables) are called coeffects
+and the output effects.
+In short, all event handlers are pure functions from coeffect map and event vector
+(containing the event name and parameters)
+to an effect map.
+
+[The listed reasons](https://github.com/Day8/re-frame/blob/master/docs/EffectfulHandlers.md#bad-why) are
+
+> 1. Cognitive load for the function's later readers goes up because they can no longer reason locally.
+> 2. Testing becomes more difficult and involves "mocking". How do we test that the http GET above is using the right URL? "mocking" should be mocked. It is a bad omen.
+> 3. And event replay-ability is lost.
+
+As REMS is a relatively simple SPA, we have found that the added indirection actually causes more cognitive load,
+we are not unit testing event handlers anyway,
+and we haven't used time travelling debugger or anything else that benefits from replay-ability.
+
+Therefore we have decided that it's ok to cause simple side-effects,
+such as fetching data from the backend,
+directly within event handlers.
+
+In other words, this is how we used to do things:
+
+```clojure
+(rf/reg-event-fx
+ ::send-request-decision
+ (fn [{{::keys [deciders application-id comment on-finished]} :db} _]
+   {:request-decision-effect [deciders application-id comment on-finished]}))
+
+(rf/reg-fx
+ :request-decision-effect
+ (fn [[deciders application-id comment on-finished]]
+   (fetch "/api/applications/request-decision"
+          {:params {:deciders deciders
+                    :application-id application-id
+                    :comment comment}
+           :handler on-finished})))
+```
+
+In some cases the effect handler was not even calling fetch directly,
+but had one more indirection of calling a specific function `fetch-this-one-thing!` without parameters.
+It's slightly better to inline the actual fetch to the effect,
+but we will instead write
+
+```clojure
+(rf/reg-event-fx
+ ::send-request-decision
+ (fn [{{::keys [deciders application-id comment on-finished]} :db} _]
+   (fetch "/api/applications/request-decision"
+          {:params {:deciders deciders
+                    :application-id application-id
+                    :comment comment}
+           :handler on-finished})
+   {}))
+```
+
+In particular do not mix side-effects and effect handlers in the same flow!
+
+## Handling user interaction in components
+
+We could also launch the fetch directly from the components, like buttons,
```

Comment:
how about "directly from the components (e.g. on-click callbacks of buttons)" or something that mentions callbacks